File: src/main/java/ch/mare/java/nullchecks/MailSender.java
package ch.mare.java.nullchecks;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

public class MailSender {

    public void sendMessageToClient(@Nullable Client client, @Nullable String message, @NotNull Mailer mailer) {
        if (client == null || message == null) {
            return;
        }

        PersonalInfo personalInfo = client.getPersonalInfo();
        if (personalInfo == null) {
            return;
        }

        String email = personalInfo.getEmail();
        if (email == null) {
            return;
        }

        mailer.sendMessage(email, message);
    }

}

class Client {

    public PersonalInfo getPersonalInfo() {
        return null;
    }
}

class Mailer {

    public void sendMessage(String email, String message) {
        System.out.println("send message " + message + " to " + email);
    }
}

class PersonalInfo {

    public String getEmail() {
        return null;
    }
}

File: src/main/kotlin/ch/mare/kotlin/nullchecks/MailSender.kt
package ch.mare.kotlin.nullchecks

fun sendMessageToClient(client: Client?, message: String?, mailer: Mailer) {
    val email = client?.personalInfo?.email
    if (email != null && message != null) {
        mailer.sendMessage(email, message)
    }

//    if(message != null){
//        client?.personalInfo?.email?.let { mailer.sendMessage(it, message) }
//    }
}

class Client {

    val personalInfo: PersonalInfo?
        get() = null
}

class Mailer {

    fun sendMessage(email: String, message: String) {
        println("send message $message to $email")
    }
}

class PersonalInfo {

    val email: String?
        get() = null
}